File: 463. Island Perimeter.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size(), ans = 0;
        vector<pair<int, int>> dirs = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
        
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {   
                if(grid[i][j]) {
                    for(auto &dir : dirs) {
                        int dx = i + dir.first, dy = j + dir.second;
                        
                        if(dx < 0 or dx >= m or dy < 0 or dy >= n or grid[dx][dy] == 0)
                            ans++;
                    }
                }
            }
        }
        
        return ans;
    }
};

File: 463. Island Perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        ans = 0

        for i in range(m):
            for j in range(n):
                if(grid[i][j] == 1):
                    if(i - 1 >= 0 and grid[i - 1][j] == 0):
                        ans += 1
                    if(i + 1 < m and grid[i + 1][j] == 0):
                        ans += 1
                    if(j - 1 >= 0 and grid[i][j - 1] == 0):
                        ans += 1
                    if(j + 1 < n and grid[i][j + 1] == 0):
                        ans += 1
                    if(i - 1 < 0):
                        ans += 1
                    if(i + 1 >= m):
                        ans += 1
                    if(j - 1 < 0):
                        ans += 1
                    if(j + 1 >= n):
                        ans += 1
        
        return ans
